Associate each dataset with the feed it validated against (Issue 110)

ct-parker opened the pull request:
## What does this do

Stores the version of the schema that each record validated against during data import.

First commit:
- add `schema_version` column to datasets table (`models/dataset.model.js`)
- add support for `schema_version` in database controller (`controllers/dataset.controller.js`)
- set `schema_version` during validation / import process (`scripts/import_datafeeds.js`)
- tested and verified with GLBRC 0.1.0 feed and other 0.0.8 feeds

Second commit:
- moved `app/config/schema` to `app/schemas` and created `index.js` to contain code related to `schema_list.json` (formerly `schema_index.json`
- added `supported` boolean to `schema_list.json` to indicate which of the schema versions are supported by the UI
- added scope to all `Dataset` queries. Specified `defaultScope` for unscoped queries, and `supportedOnly` to indicate those that should be restricted to schema versions supported by the UI.
- tested and verified with GLBRC 0.1.0 feed and other 0.0.8 feeds (see screenshots)

## Related Issues

Fixes #110 

## Screenshots

Using `https://fair-data.glbrc.org/glbrc-0.1.0.json` as the GLBRC data feed:

```
C:\Progra~1\PostgreSQL\11\bin\psql.exe -U postgres -p 6432

postgres=# SELECT * FROM datasets LIMIT 1;
       uid        | schema_version |                                                                                                                                                                                                                                                                                                                                                       json                                                                                                                                                                                                                                                                                                                                                        |         createdAt         |         updatedAt
------------------+----------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------------+---------------------------
 JBEI_JPUB_000267 | 0.0.8          | {"brc": "JBEI", "date": "2011-08-09", "title": "pBbE0c-RFP", "creator": [{"email": "tslee@lbl.gov", "affiliation": "Joint BioEnergy Institute, Lawrence Berkeley National Laboratory Berkeley CA 94720", "creatorName": "Taek Soon Lee", "primaryContact": true}], "keywords": [], "identifier": "JPUB_000267", "repository": "ICE", "datasetName": "Lee et al. 2011", "description": "BglBrick vector based on the work of Rachel Krupa w/constitutive RFP cassette\nconstitutive RFP cassette between BglBrick sites for easy identification of background colonies\nSelection Markers: Chloramphenicol\nOrigin of Replication: colE1,", "bibliographicCitation": "https://public-registry.jbei.org/folders/1"} | 2025-04-17 11:39:09.65+00 | 2025-04-17 11:39:09.65+00
(1 row)

postgres=# SELECT schema_version,COUNT(uid) FROM datasets GROUP BY schema_version;
 schema_version | count
----------------+-------
 0.1.0          |    37
 0.0.8          | 11470
(2 rows)

postgres=#
```

When `schema_list.json` has only version 0.0.8 marked as supported and the GLBRC 0.1.0 feed is imported, no GLBRC records show up in the UI or API:

![v008_supported_only](https://github.com/user-attachments/assets/0c12eaa6-53a3-412d-a7be-67197fd4a823)

When `schema_list.json` also has version 0.1.0 marked as supported and the GLBRC 0.1.0 feed is imported, GLBRC records are shown in both the UI and API:

![v008-010_supported](https://github.com/user-attachments/assets/b40cf921-029e-4568-bdea-46e4a09393d0)

## Acceptance

Add an 'x' to the boxes below if they are complete
- [x] My changes work as expected locally
- [x] My changes are related to existing issues or other approved work
- [ ] I updated the Changelog (if applicable)
- [ ] I added tests with appropriate coverage and verified they all pass (if applicable)
- [ ] I updated user documentation (if applicable)

ct-parker commented:
Note: There are additional commits shown because the issue_110 branch is a sub-branch of issue_99.

throwern commented:
After moving the schema files to a new folder structure, I think we need to cleanup the old files. It looks like everything under `api/app/config/schema/` should be removed since those files are are now in `api/app/schemas/`

ct-parker commented:
Adding discussion from Slack channel between Nick and Charles from Friday April 18:

Nick:

- Do we need the default scope applied? I thought this was automatic/default and only required if we wanted multiple scopes and including `defaultScope`.
- Would you be okay changing the name from `supportedOnly` to something like `supportedVersions` for a bit more clarity?
- There seems to be two different schema folders now: `api/app/config/schema` and `api/app/schemas`. I'm thinking we need to git delete the folder we don't need anymore.

Charles:
- Regarding the default scope, I feel that this should be specified (even if not needed) in order to indicate to other developers that those specific queries are supposed to use the `defaultScope` (so it doesn't look like an accidental omission).
- Regarding the `supportedVersions` scope - yes, that sounds good to me. I was following the naming conventions from the documentation example, but your recommendation makes more sense.
- Correct, there should only be: `api/app/schemas`.

Nick:
- Did you happen to test from a fresh DB? My local db with prior data failed to sync the new column. I can fix it by forcing a full sync (drop and recreate all tables): `db.sequelize.sync({force: true}); `but not sure we want to commit that for prod. I think the suggested setup from sequelize is using migrations. Clint and I are used to doing that with Ruby on Rails apps but it's more boilerplate to get setup.

Charles:
- There is also: `db.sequelize.sync({ alter: true };` see [documentation](https://sequelize.org/docs/v6/core-concepts/model-basics/#model-synchronization). This checks what is the current state of the table in the database (which columns it has, what are their data types, etc.), and then performs the necessary changes in the table to make it match the model. But `alter:true` will cause a different problem...with this specific change `schema_version` is not allowed to be `null`, so the table change will fail unless we can provide a default value of 0.0.8 via `dataset.model.js`. This would be possible by adding `defaultValue: '0.0.8'`
to `schema_version`.
- Here is my proposal:
  - In `dataset.model.js` add `, defaultValue: '0.0.8'` for `schema_version`.
  - In `server.js` change `db.sequelize.sync();` to `await db.sequelize.sync({ alter: true });`.

Nick:
- I did test the `alter: true` and ran into the issue with `not null`. Setting a default value works for me and does seem like a small change.
- The 'migrations' option gives a nice way to control this. You get to create separate files specifically for DDL. Then you run commands during deploy to 'apply' those changes like `db:migrate` and status of what has/has not been applied is tracked.  The separate management avoids issues with timing in app startup. I'm not sure we want to set it up right now though...maybe a workaround for this PR and discuss enabling migrations at the next meeting?

Charles:
- Agreed.

throwern commented on api/app/models/dataset.model.js:
```diff
@@ -7,6 +7,10 @@ module.exports = (sequelize, Sequelize) => {
     },
+    schema_version: {
+      type: DataTypes.STRING,
+      allowNull: false
```
```suggestion
      allowNull: false,
      defaultValue: "0.0.8"
```

throwern reviewed (commented):
I verified that `alter: true` works after setting a default value for the new column. I added the suggested model change but I cannot suggest the sync change without a nearby diff: https://github.com/orgs/community/discussions/59057

I did find notes in the documentation indicating that we should consider other options for production use:
- https://sequelize.org/api/v6/class/src/sequelize.js~sequelize#instance-method-sync

> Alters tables to fit models. Provide an object for additional configuration. Not recommended for production use. If not further configured deletes data in columns that were removed or had their type changed in the model.

This comment in the sequelize project suggests including `drop: false`, which seems to be what we want and is what I tested:
- https://github.com/sequelize/sequelize/issues/537#issuecomment-1073194298
> Prevents any drop statements while altering a table when set to false

I think we can use `db.sequelize.sync({ alter: { drop: false } });` in our `server.js` until we decide to use migrations.

ct-parker commented:
> I verified that `alter: true` works after setting a default value for the new column. I added the suggested model change but I cannot suggest the sync change without a nearby diff.
> ...
> I think we can use `db.sequelize.sync({ alter: { drop: false } });` in our `server.js` until we decide to use migrations.

I added a commit for this to this PR. Although, please note this is not tested. I am not currently set up to test this behavior.
